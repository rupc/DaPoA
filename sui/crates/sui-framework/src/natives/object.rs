// Copyright (c) Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

use crate::{legacy_emit_cost, natives::object_runtime::ObjectRuntime};
use move_binary_format::errors::PartialVMResult;
use move_core_types::account_address::AccountAddress;
use move_vm_runtime::native_functions::NativeContext;
use move_vm_types::{
    loaded_data::runtime_types::Type,
    natives::function::NativeResult,
    pop_arg,
    values::{StructRef, Value},
};
use smallvec::smallvec;
use std::collections::VecDeque;

// native fun borrow_uid<T: key>(obj: &T): &UID;
pub fn borrow_uid(
    _context: &mut NativeContext,
    ty_args: Vec<Type>,
    mut args: VecDeque<Value>,
) -> PartialVMResult<NativeResult> {
    debug_assert!(ty_args.len() == 1);
    debug_assert!(args.len() == 1);

    let obj = pop_arg!(args, StructRef);
    let id_field = obj.borrow_field(0)?;

    // TODO: what should the cost of this be?
    let cost = legacy_emit_cost();

    Ok(NativeResult::ok(cost, smallvec![id_field]))
}

// native fun delete_impl(id: address);
pub fn delete_impl(
    context: &mut NativeContext,
    ty_args: Vec<Type>,
    mut args: VecDeque<Value>,
) -> PartialVMResult<NativeResult> {
    debug_assert!(ty_args.is_empty());
    debug_assert!(args.len() == 1);

    // unwrap safe because the interface of native function guarantees it.
    let uid_bytes = pop_arg!(args, AccountAddress);

    // TODO: what should the cost of this be?
    let cost = legacy_emit_cost();

    let obj_runtime: &mut ObjectRuntime = context.extensions_mut().get_mut();
    obj_runtime.delete_id(uid_bytes.into())?;
    Ok(NativeResult::ok(cost, smallvec![]))
}

// native fun record_new_uid(id: address);
pub fn record_new_uid(
    context: &mut NativeContext,
    ty_args: Vec<Type>,
    mut args: VecDeque<Value>,
) -> PartialVMResult<NativeResult> {
    debug_assert!(ty_args.is_empty());
    debug_assert!(args.len() == 1);
    // unwrap safe because the interface of native function guarantees it.
    let uid_bytes = pop_arg!(args, AccountAddress);

    // TODO: what should the cost of this be?
    let cost = legacy_emit_cost();

    let obj_runtime: &mut ObjectRuntime = context.extensions_mut().get_mut();
    obj_runtime.new_id(uid_bytes.into())?;
    Ok(NativeResult::ok(cost, smallvec![]))
}
