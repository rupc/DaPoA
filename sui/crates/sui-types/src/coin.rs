// Copyright (c) Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

use move_core_types::{
    ident_str,
    identifier::IdentStr,
    language_storage::{StructTag, TypeTag},
    value::{MoveFieldLayout, MoveStructLayout, MoveTypeLayout},
};
use serde::{Deserialize, Serialize};

use crate::base_types::SequenceNumber;
use crate::committee::EpochId;
use crate::object::{MoveObject, Owner};
use crate::storage::{DeleteKind, SingleTxContext, WriteKind};
use crate::temporary_store::TemporaryStore;
use crate::{
    balance::{Balance, Supply},
    error::{ExecutionError, ExecutionErrorKind},
    object::{Data, Object},
};
use crate::{base_types::TransactionDigest, error::SuiError};
use crate::{
    base_types::{ObjectID, SuiAddress},
    id::UID,
    SUI_FRAMEWORK_ADDRESS,
};
use schemars::JsonSchema;

pub const COIN_MODULE_NAME: &IdentStr = ident_str!("coin");
pub const COIN_STRUCT_NAME: &IdentStr = ident_str!("Coin");
pub const COIN_METADATA_STRUCT_NAME: &IdentStr = ident_str!("CoinMetadata");
pub const COIN_TREASURE_CAP_NAME: &IdentStr = ident_str!("TreasuryCap");

pub const PAY_MODULE_NAME: &IdentStr = ident_str!("pay");
pub const PAY_JOIN_FUNC_NAME: &IdentStr = ident_str!("join");
pub const PAY_SPLIT_N_FUNC_NAME: &IdentStr = ident_str!("divide_and_keep");
pub const PAY_SPLIT_VEC_FUNC_NAME: &IdentStr = ident_str!("split_vec");

pub const LOCKED_COIN_MODULE_NAME: &IdentStr = ident_str!("locked_coin");
pub const LOCKED_COIN_STRUCT_NAME: &IdentStr = ident_str!("LockedCoin");

// Rust version of the Move sui::coin::Coin type
#[derive(Debug, Serialize, Deserialize, Clone, JsonSchema, Eq, PartialEq)]
pub struct Coin {
    pub id: UID,
    pub balance: Balance,
}

impl Coin {
    pub fn new(id: UID, value: u64) -> Self {
        Self {
            id,
            balance: Balance::new(value),
        }
    }

    pub fn type_(type_param: StructTag) -> StructTag {
        StructTag {
            address: SUI_FRAMEWORK_ADDRESS,
            name: COIN_STRUCT_NAME.to_owned(),
            module: COIN_MODULE_NAME.to_owned(),
            type_params: vec![TypeTag::Struct(Box::new(type_param))],
        }
    }

    /// Is this other StructTag representing a Coin?
    pub fn is_coin(other: &StructTag) -> bool {
        other.address == SUI_FRAMEWORK_ADDRESS
            && other.module.as_ident_str() == COIN_MODULE_NAME
            && other.name.as_ident_str() == COIN_STRUCT_NAME
    }

    /// Create a coin from BCS bytes
    pub fn from_bcs_bytes(content: &[u8]) -> Result<Self, ExecutionError> {
        bcs::from_bytes(content).map_err(|err| {
            ExecutionError::new_with_source(
                ExecutionErrorKind::InvalidCoinObject,
                format!("Unable to deserialize coin object: {:?}", err),
            )
        })
    }

    /// If the given object is a Coin, deserialize its contents and extract the balance Ok(Some(u64)).
    /// If it's not a Coin, return Ok(None).
    /// The cost is 2 comparisons if not a coin, and deserialization if its a Coin.
    pub fn extract_balance_if_coin(object: &Object) -> Result<Option<u64>, ExecutionError> {
        match &object.data {
            Data::Move(move_obj) => {
                if !Self::is_coin(&move_obj.type_) {
                    return Ok(None);
                }

                let coin = Self::from_bcs_bytes(move_obj.contents())?;
                Ok(Some(coin.value()))
            }
            _ => Ok(None), // package
        }
    }

    pub fn id(&self) -> &ObjectID {
        self.id.object_id()
    }

    pub fn value(&self) -> u64 {
        self.balance.value()
    }

    pub fn to_bcs_bytes(&self) -> Vec<u8> {
        bcs::to_bytes(&self).unwrap()
    }

    pub fn layout(type_param: StructTag) -> MoveStructLayout {
        MoveStructLayout::WithTypes {
            type_: Self::type_(type_param.clone()),
            fields: vec![
                MoveFieldLayout::new(
                    ident_str!("id").to_owned(),
                    MoveTypeLayout::Struct(UID::layout()),
                ),
                MoveFieldLayout::new(
                    ident_str!("balance").to_owned(),
                    MoveTypeLayout::Struct(Balance::layout(type_param)),
                ),
            ],
        }
    }

    // Shift balance of coins_to_merge to this coin.
    // Related coin objects need to be updated in temporary_store to persist the changes,
    // including deleting the coin objects that have been merged.
    pub fn merge_coins(&mut self, coins_to_merge: &mut [Coin]) -> Result<(), ExecutionError> {
        let Some(total_coins) =
            coins_to_merge.iter().fold(Some(0u64), |acc, c| acc?.checked_add(c.value())) else {
                return Err(ExecutionError::new_with_source(
                    ExecutionErrorKind::CoinTooLarge,
                    format!("Coin {} exceeds maximum value", self.id())
                ))
            };

        for coin in coins_to_merge.iter_mut() {
            // unwrap() is safe because balance value is the same as coin value
            coin.balance.withdraw(coin.value()).unwrap();
        }
        let Some(new_balance) = self.value().checked_add(total_coins) else {
            return Err(ExecutionError::new_with_source(
                ExecutionErrorKind::CoinTooLarge,
                format!("Coin {} exceeds maximum value", self.id())
            ))
        };
        self.balance = Balance::new(new_balance);
        Ok(())
    }

    // Split amount out of this coin to a new coin.
    // Related coin objects need to be updated in temporary_store to persist the changes,
    // including creating the coin object related to the newly created coin.
    pub fn split_coin(&mut self, amount: u64, new_coin_id: UID) -> Result<Coin, ExecutionError> {
        self.balance.withdraw(amount)?;
        Ok(Coin::new(new_coin_id, amount))
    }
}

// Rust version of the Move sui::coin::TreasuryCap type
#[derive(Debug, Serialize, Deserialize, Clone, Eq, PartialEq, JsonSchema)]
pub struct TreasuryCap {
    pub id: UID,
    pub total_supply: Supply,
}

impl TreasuryCap {
    /// Create a TreasuryCap from BCS bytes
    pub fn from_bcs_bytes(content: &[u8]) -> Result<Self, SuiError> {
        bcs::from_bytes(content).map_err(|err| SuiError::TypeError {
            error: format!("Unable to deserialize TreasuryCap object: {:?}", err),
        })
    }

    pub fn type_(type_param: StructTag) -> StructTag {
        StructTag {
            address: SUI_FRAMEWORK_ADDRESS,
            name: COIN_TREASURE_CAP_NAME.to_owned(),
            module: COIN_MODULE_NAME.to_owned(),
            type_params: vec![TypeTag::Struct(Box::new(type_param))],
        }
    }
}

pub fn transfer_coin<S>(
    ctx: &SingleTxContext,
    temporary_store: &mut TemporaryStore<S>,
    coin: &Coin,
    recipient: SuiAddress,
    coin_type: StructTag,
    previous_transaction: TransactionDigest,
) {
    let new_coin = Object::new_move(
        MoveObject::new_coin(coin_type, SequenceNumber::new(), *coin.id(), coin.value()),
        Owner::AddressOwner(recipient),
        previous_transaction,
    );
    temporary_store.write_object(ctx, new_coin, WriteKind::Create);
}

// A helper function for pay_sui and pay_all_sui.
// It updates the gas_coin_obj based on the updated gas_coin, transfers gas_coin_obj to
// recipient when needed, and then deletes all other input coins other than gas_coin_obj.
pub fn update_input_coins<S>(
    ctx: &SingleTxContext,
    temporary_store: &mut TemporaryStore<S>,
    coin_objects: &mut Vec<Object>,
    gas_coin: &Coin,
    recipient: Option<SuiAddress>,
) {
    let mut gas_coin_obj = coin_objects.remove(0);
    let new_contents = bcs::to_bytes(gas_coin).expect("Coin serialization should not fail");
    // unwrap is safe because we checked that it was a coin object above.
    let move_obj = gas_coin_obj.data.try_as_move_mut().unwrap();
    move_obj.update_coin_contents(new_contents);
    if let Some(recipient) = recipient {
        gas_coin_obj.transfer(recipient);
    }
    temporary_store.write_object(ctx, gas_coin_obj, WriteKind::Mutate);

    for coin_object in coin_objects.iter() {
        temporary_store.delete_object(
            ctx,
            &coin_object.id(),
            coin_object.version(),
            DeleteKind::Normal,
        )
    }
}

// Rust version of the Move sui::coin::CoinMetadata type
#[derive(Debug, Serialize, Deserialize, Clone, JsonSchema, Eq, PartialEq)]
pub struct CoinMetadata {
    pub id: UID,
    /// Number of decimal places the coin uses.
    pub decimals: u8,
    /// Name for the token
    pub name: String,
    /// Symbol for the token
    pub symbol: String,
    /// Description of the token
    pub description: String,
    /// URL for the token logo
    pub icon_url: Option<String>,
}

impl CoinMetadata {
    /// Is this other StructTag representing a CoinMetadata?
    pub fn is_coin_metadata(other: &StructTag) -> bool {
        other.address == SUI_FRAMEWORK_ADDRESS
            && other.module.as_ident_str() == COIN_MODULE_NAME
            && other.name.as_ident_str() == COIN_METADATA_STRUCT_NAME
    }

    /// Create a coin from BCS bytes
    pub fn from_bcs_bytes(content: &[u8]) -> Result<Self, SuiError> {
        bcs::from_bytes(content).map_err(|err| SuiError::TypeError {
            error: format!("Unable to deserialize CoinMetadata object: {:?}", err),
        })
    }

    pub fn type_(type_param: StructTag) -> StructTag {
        StructTag {
            address: SUI_FRAMEWORK_ADDRESS,
            name: COIN_METADATA_STRUCT_NAME.to_owned(),
            module: COIN_MODULE_NAME.to_owned(),
            type_params: vec![TypeTag::Struct(Box::new(type_param))],
        }
    }
}

impl TryFrom<Object> for CoinMetadata {
    type Error = SuiError;
    fn try_from(object: Object) -> Result<Self, Self::Error> {
        match &object.data {
            Data::Move(o) => {
                if CoinMetadata::is_coin_metadata(&o.type_) {
                    return CoinMetadata::from_bcs_bytes(o.contents());
                }
            }
            Data::Package(_) => {}
        }

        Err(SuiError::TypeError {
            error: format!("Object type is not a CoinMetadata: {:?}", object),
        })
    }
}
// Rust version of the Move sui::locked_coin::LockedCoin type
#[derive(Debug, Serialize, Deserialize, Clone, JsonSchema, Eq, PartialEq)]
pub struct LockedCoin {
    pub id: UID,
    pub balance: Balance,
    pub locked_until_epoch: EpochId,
}

impl LockedCoin {
    /// Is this other StructTag representing a Coin?
    pub fn is_locked_coin(other: &StructTag) -> bool {
        other.address == SUI_FRAMEWORK_ADDRESS
            && other.module.as_ident_str() == LOCKED_COIN_MODULE_NAME
            && other.name.as_ident_str() == LOCKED_COIN_STRUCT_NAME
    }
}
