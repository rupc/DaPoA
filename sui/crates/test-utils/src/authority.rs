// Copyright (c) Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

use mysten_metrics::RegistryService;
use prometheus::Registry;
use rand::{prelude::StdRng, SeedableRng};
use std::net::IpAddr;
use std::time::Duration;
use sui_config::{builder::ConfigBuilder, NetworkConfig, NodeConfig, ValidatorInfo};
use sui_core::authority_client::AuthorityAPI;
use sui_core::authority_client::NetworkAuthorityClient;
pub use sui_node::{SuiNode, SuiNodeHandle};
use sui_types::base_types::ObjectID;
use sui_types::crypto::TEST_COMMITTEE_SIZE;
use sui_types::messages::ObjectInfoRequest;
use sui_types::object::Object;

/// The default network buffer size of a test authority.
pub const NETWORK_BUFFER_SIZE: usize = 65_000;

/// Make an authority config for each of the `TEST_COMMITTEE_SIZE` authorities in the test committee.
pub fn test_authority_configs() -> NetworkConfig {
    test_and_configure_authority_configs(TEST_COMMITTEE_SIZE)
}

pub fn test_and_configure_authority_configs(committee_size: usize) -> NetworkConfig {
    let config_dir = tempfile::tempdir().unwrap().into_path();
    let rng = StdRng::from_seed([0; 32]);
    let mut configs = NetworkConfig::generate_with_rng(&config_dir, committee_size, rng);
    for config in configs.validator_configs.iter_mut() {
        let parameters = &mut config.consensus_config.as_mut().unwrap().narwhal_config;
        // NOTE: the following parameters are important to ensure tests run fast. Using the default
        // Narwhal parameters may result in tests taking >60 seconds.
        parameters.header_num_of_batches_threshold = 1;
        parameters.max_header_delay = Duration::from_millis(200);
        parameters.min_header_delay = Duration::from_millis(200);
        parameters.batch_size = 1;
        parameters.max_batch_delay = Duration::from_millis(200);
    }
    configs
}

pub fn test_authority_configs_with_objects(objects: Vec<Object>) -> (NetworkConfig, Vec<Object>) {
    let config_dir = tempfile::tempdir().unwrap().into_path();
    let rng = StdRng::from_seed([0; 32]);
    let mut configs = ConfigBuilder::new(&config_dir)
        .rng(rng)
        .committee_size(TEST_COMMITTEE_SIZE.try_into().unwrap())
        .with_objects(objects.clone())
        .build();

    for config in configs.validator_configs.iter_mut() {
        let parameters = &mut config.consensus_config.as_mut().unwrap().narwhal_config;
        // NOTE: the following parameters are important to ensure tests run fast. Using the default
        // Narwhal parameters may result in tests taking >60 seconds.
        parameters.header_num_of_batches_threshold = 1;
        parameters.max_header_delay = Duration::from_millis(200);
        parameters.min_header_delay = Duration::from_millis(200);
        parameters.batch_size = 1;
        parameters.max_batch_delay = Duration::from_millis(200);
    }

    let objects = objects
        .into_iter()
        .map(|o| configs.genesis.object(o.id()).unwrap())
        .collect();

    (configs, objects)
}

#[cfg(not(msim))]
pub async fn start_node(config: &NodeConfig, registry_service: RegistryService) -> SuiNodeHandle {
    SuiNode::start(config, registry_service)
        .await
        .unwrap()
        .into()
}

/// In the simulator, we call SuiNode::start from inside a newly spawned simulator node.
/// However, we then immediately return the SuiNode handle back to the caller. The caller now has
/// a direct handle to an object that is "running" on a different "machine". By itself, this
/// doesn't break anything in the simulator, it just allows test code to magically mutate state
/// that is owned by some other machine.
///
/// Most of the time, tests do this just in order to verify some internal state, so this is fine
/// most of the time.
#[cfg(msim)]
pub async fn start_node(config: &NodeConfig, registry_service: RegistryService) -> SuiNodeHandle {
    use std::net::SocketAddr;

    let config = config.clone();
    let socket_addr = mysten_network::multiaddr::to_socket_addr(&config.network_address).unwrap();
    let ip = match socket_addr {
        SocketAddr::V4(v4) => IpAddr::V4(*v4.ip()),
        _ => panic!("unsupported protocol"),
    };

    let handle = sui_simulator::runtime::Handle::current();
    let builder = handle.create_node();
    tracing::info!("starting new node with ip {:?}", ip);
    let node = builder
        .ip(ip)
        .name(format!("{:?}", config.protocol_public_key().concise()))
        .init(|| async {
            tracing::info!("node restarted");
        })
        .build();

    node.spawn(async move { SuiNode::start(&config, registry_service).await.unwrap() })
        .await
        .unwrap()
        .into()
}

/// Spawn all authorities in the test committee into a separate tokio task.
pub async fn spawn_test_authorities<I>(objects: I, config: &NetworkConfig) -> Vec<SuiNodeHandle>
where
    I: IntoIterator<Item = Object> + Clone,
{
    let mut handles = Vec::new();
    for validator in config.validator_configs() {
        let registry_service = RegistryService::new(Registry::new());
        let node = start_node(validator, registry_service).await;
        let objects = objects.clone();

        node.with_async(|node| async move {
            let state = node.state();
            for o in objects {
                state.insert_genesis_object(o).await
            }
        })
        .await;

        handles.push(node);
    }
    handles
}

/// This function can be called after `spawn_test_authorities` to
/// start a fullnode.
pub async fn spawn_fullnode(config: &NetworkConfig, rpc_port: Option<u16>) -> SuiNodeHandle {
    let registry_service = RegistryService::new(Registry::new());

    let mut builder = config.fullnode_config_builder();

    if cfg!(msim) {
        let ip_addr: IpAddr = "11.10.0.0".to_string().parse().unwrap();
        builder = builder
            .with_listen_ip(ip_addr)
            .with_port(8080)
            .with_p2p_port(8084)
            .with_rpc_port(rpc_port.unwrap_or(9000))
            .with_admin_port(8888);
    } else {
        builder = builder.set_rpc_port(rpc_port);
    }

    let fullnode_config = builder.build().unwrap();
    start_node(&fullnode_config, registry_service).await
}

/// Get a network client to communicate with the consensus.
pub fn get_client(config: &ValidatorInfo) -> NetworkAuthorityClient {
    NetworkAuthorityClient::connect_lazy(config.network_address()).unwrap()
}

pub async fn get_object(config: &ValidatorInfo, object_id: ObjectID) -> Object {
    get_client(config)
        .handle_object_info_request(ObjectInfoRequest::latest_object_info_request(
            object_id, None,
        ))
        .await
        .unwrap()
        .object
}
