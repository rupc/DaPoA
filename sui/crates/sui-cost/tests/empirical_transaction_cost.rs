// Copyright (c) Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

use insta::assert_json_snapshot;
use std::{collections::BTreeMap, path::PathBuf};
use sui_config::NetworkConfig;
use sui_core::test_utils::make_transfer_object_transaction;
use sui_core::test_utils::make_transfer_sui_transaction;
use sui_types::base_types::SuiAddress;
use sui_types::coin::PAY_JOIN_FUNC_NAME;
use sui_types::coin::PAY_MODULE_NAME;
use sui_types::coin::PAY_SPLIT_VEC_FUNC_NAME;
use sui_types::crypto::{deterministic_random_account_key, AccountKeyPair};
use sui_types::messages::VerifiedTransaction;
use sui_types::object::{generate_test_gas_objects, Object};
use sui_types::SUI_FRAMEWORK_OBJECT_ID;
use sui_types::{
    gas::GasCostSummary,
    messages::{CallArg, ExecutionStatus, ObjectArg},
};
use test_utils::authority::spawn_test_authorities;
use test_utils::messages::move_transaction_with_type_tags;
use test_utils::transaction::make_publish_package;
use test_utils::{
    authority::test_authority_configs,
    messages::move_transaction,
    transaction::{
        publish_counter_package, submit_shared_object_transaction, submit_single_owner_transaction,
    },
};
// Copyright (c) Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

use serde::{Deserialize, Serialize};
use strum_macros::Display;
use strum_macros::EnumString;

#[derive(
    Debug, Eq, PartialEq, Hash, Serialize, Deserialize, Ord, PartialOrd, Clone, Display, EnumString,
)]
pub enum CommonTransactionCosts {
    Publish,
    MergeCoin,
    SplitCoin(usize),
    TransferWholeCoin,
    TransferWholeSuiCoin,
    TransferPortionSuiCoin,
    SharedCounterCreate,
    SharedCounterAssertValue,
    SharedCounterIncrement,
}

impl CommonTransactionCosts {
    pub fn is_shared_object_tx(&self) -> bool {
        matches!(
            self,
            CommonTransactionCosts::SharedCounterAssertValue
                | CommonTransactionCosts::SharedCounterIncrement
        )
    }
}

const TEST_DATA_DIR: &str = "tests/data/";

// Execute every entry function in Move framework and examples and ensure costs don't change
// To review snapshot changes, and fix snapshot differences,
// 0. Install cargo-insta
// 1. Run `cargo insta test --review` under `./sui-cost`.
// 2. Review, accept or reject changes.

#[tokio::test]
async fn test_good_snapshot() -> Result<(), anyhow::Error> {
    let mut common_costs_actual: BTreeMap<String, GasCostSummary> = BTreeMap::new();

    run_actual_costs().await?.iter().for_each(|(k, actual)| {
        common_costs_actual.insert(k.to_string(), actual.clone());
    });
    assert_json_snapshot!(common_costs_actual);

    Ok(())
}

async fn split_n_tx(n: u64, coin: &Object, gas: &Object) -> VerifiedTransaction {
    let split_amounts = vec![10u64; n as usize];
    let type_args = vec![coin.get_move_template_type().unwrap()];

    move_transaction_with_type_tags(
        gas.clone(),
        PAY_MODULE_NAME.as_str(),
        PAY_SPLIT_VEC_FUNC_NAME.as_str(),
        SUI_FRAMEWORK_OBJECT_ID,
        &type_args,
        vec![
            CallArg::Object(ObjectArg::ImmOrOwnedObject(coin.compute_object_reference())),
            CallArg::Pure(bcs::to_bytes(&split_amounts).unwrap()),
        ],
    )
}

async fn create_txes(
    sender: SuiAddress,
    keypair: &AccountKeyPair,
    gas_objects: &[Object],
    configs: &NetworkConfig,
) -> BTreeMap<CommonTransactionCosts, VerifiedTransaction> {
    let mut ret = BTreeMap::new();
    let mut gas_objects = gas_objects.to_vec().clone();
    // let _handles = spawn_test_authorities(gas_objects.clone(), configs).await;
    // Publish the move package to all authorities and get the new package ref.
    tokio::task::yield_now().await;
    tokio::time::sleep(std::time::Duration::from_secs(3)).await;

    //
    // Publish
    //
    let mut package_path = PathBuf::from(TEST_DATA_DIR);
    package_path.push("dummy_modules_publish");
    let publish_tx = make_publish_package(gas_objects.pop().unwrap(), package_path);
    ret.insert(CommonTransactionCosts::Publish, publish_tx);

    //
    // Transfer Whole Sui Coin and Transfer Portion of Sui Coin
    //
    let whole_sui_coin_tx = make_transfer_sui_transaction(
        gas_objects.pop().unwrap().compute_object_reference(),
        SuiAddress::default(),
        None,
        sender,
        keypair,
        None,
    );
    let partial_sui_coin_tx = make_transfer_sui_transaction(
        gas_objects.pop().unwrap().compute_object_reference(),
        SuiAddress::default(),
        Some(100),
        sender,
        keypair,
        None,
    );
    ret.insert(
        CommonTransactionCosts::TransferWholeSuiCoin,
        whole_sui_coin_tx,
    );
    ret.insert(
        CommonTransactionCosts::TransferPortionSuiCoin,
        partial_sui_coin_tx,
    );

    //
    // Transfer Whole Coin Object
    //
    let whole_coin_tx = make_transfer_object_transaction(
        gas_objects.pop().unwrap().compute_object_reference(),
        gas_objects.pop().unwrap().compute_object_reference(),
        sender,
        keypair,
        SuiAddress::default(),
        None,
    );

    ret.insert(CommonTransactionCosts::TransferWholeCoin, whole_coin_tx);

    //
    // Merge Two Coins
    //
    let c1 = gas_objects.pop().unwrap();
    let type_args = vec![c1.get_move_template_type().unwrap()];

    let merge_tx = move_transaction_with_type_tags(
        gas_objects.pop().unwrap(),
        PAY_MODULE_NAME.as_str(),
        PAY_JOIN_FUNC_NAME.as_str(),
        SUI_FRAMEWORK_OBJECT_ID,
        &type_args,
        vec![
            CallArg::Object(ObjectArg::ImmOrOwnedObject(c1.compute_object_reference())),
            CallArg::Object(ObjectArg::ImmOrOwnedObject(
                gas_objects.pop().unwrap().compute_object_reference(),
            )),
        ],
    );
    ret.insert(CommonTransactionCosts::MergeCoin, merge_tx);

    //
    // Splt A Coin Into N Specific Amounts
    // Note spltting complexity does not depend on the amounts but only on the number of amounts
    //
    for n in 0..4 {
        let gas = gas_objects.pop().unwrap();
        let coin = gas_objects.pop().unwrap();
        let split_tx = split_n_tx(n, &gas, &coin).await.clone();
        ret.insert(CommonTransactionCosts::SplitCoin(n as usize), split_tx);
    }

    //
    // Shared Object Section
    // Using the `counter` example
    //

    let package_id = publish_counter_package(gas_objects.pop().unwrap(), &configs.validator_set())
        .await
        .0;

    // Make a transaction to create a counter.
    tokio::task::yield_now().await;
    let transaction = move_transaction(
        gas_objects.pop().unwrap(),
        "counter",
        "create",
        package_id,
        /* arguments */ Vec::default(),
    );
    let (effects, _) =
        submit_single_owner_transaction(transaction.clone(), &configs.validator_set()).await;
    assert!(matches!(effects.status, ExecutionStatus::Success { .. }));
    let ((counter_id, counter_initial_shared_version, _), _) = effects.created[0];
    let counter_object_arg = ObjectArg::SharedObject {
        id: counter_id,
        initial_shared_version: counter_initial_shared_version,
        mutable: true,
    };

    ret.insert(CommonTransactionCosts::SharedCounterCreate, transaction);

    // Ensure the value of the counter is `0`.
    tokio::task::yield_now().await;
    let transaction = move_transaction(
        gas_objects.pop().unwrap(),
        "counter",
        "assert_value",
        package_id,
        vec![
            CallArg::Object(counter_object_arg),
            CallArg::Pure(0u64.to_le_bytes().to_vec()),
        ],
    );

    ret.insert(
        CommonTransactionCosts::SharedCounterAssertValue,
        transaction,
    );

    // Make a transaction to increment the counter.
    tokio::task::yield_now().await;
    let transaction = move_transaction(
        gas_objects.pop().unwrap(),
        "counter",
        "increment",
        package_id,
        vec![CallArg::Object(counter_object_arg)],
    );

    ret.insert(CommonTransactionCosts::SharedCounterIncrement, transaction);

    ret
}

async fn run_actual_costs(
) -> Result<BTreeMap<CommonTransactionCosts, GasCostSummary>, anyhow::Error> {
    let mut ret = BTreeMap::new();
    let gas_objects = generate_test_gas_objects();
    let (sender, keypair) = deterministic_random_account_key();

    // Get the authority configs and spawn them. Note that it is important to not drop
    // the handles (or the authorities will stop).
    let configs = test_authority_configs();
    let _ = spawn_test_authorities(gas_objects.clone(), &configs).await;
    // Publish the move package to all authorities and get the new package ref.
    tokio::task::yield_now().await;
    tokio::time::sleep(std::time::Duration::from_secs(3)).await;

    let tx_map = create_txes(sender, &keypair, &gas_objects, &configs).await;
    for (tx_type, tx) in tx_map {
        let gas_used = if tx_type.is_shared_object_tx() {
            submit_shared_object_transaction(tx, &configs.validator_set())
                .await
                .unwrap()
                .0
                .gas_cost_summary()
                .clone()
        } else {
            submit_single_owner_transaction(tx, &configs.validator_set())
                .await
                .0
                .gas_cost_summary()
                .clone()
        };

        ret.insert(tx_type, gas_used);
    }
    Ok(ret)
}
