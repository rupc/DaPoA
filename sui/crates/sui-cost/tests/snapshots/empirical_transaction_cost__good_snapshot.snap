---
source: crates/sui-cost/tests/empirical_transaction_cost.rs
expression: common_costs_actual
---
{
  "MergeCoin": {
    "computation_cost": 176,
    "storage_cost": 42,
    "storage_rebate": 0
  },
  "Publish": {
    "computation_cost": 341,
    "storage_cost": 140,
    "storage_rebate": 0
  },
  "SharedCounterAssertValue": {
    "computation_cost": 212,
    "storage_cost": 42,
    "storage_rebate": 21
  },
  "SharedCounterCreate": {
    "computation_cost": 216,
    "storage_cost": 42,
    "storage_rebate": 0
  },
  "SharedCounterIncrement": {
    "computation_cost": 212,
    "storage_cost": 42,
    "storage_rebate": 21
  },
  "SplitCoin": {
    "computation_cost": 290,
    "storage_cost": 105,
    "storage_rebate": 0
  },
  "TransferPortionSuiCoin": {
    "computation_cost": 157,
    "storage_cost": 63,
    "storage_rebate": 0
  },
  "TransferWholeCoin": {
    "computation_cost": 151,
    "storage_cost": 42,
    "storage_rebate": 0
  },
  "TransferWholeSuiCoin": {
    "computation_cost": 148,
    "storage_cost": 42,
    "storage_rebate": 0
  }
}
